Knight minimum steps chess

albero94 opened the pull request.

albero94 committed:
Adding Knight minimum steps problem

albero94 committed:
Adding Knight minimum steps problem

bushidocodes commented on src/test/java/edu/gwu/csci6212/MinimumStepsChessKnightTest.java:
```diff
@@ -0,0 +1,56 @@
+public class MinimumStepsChessKnightTest {
+
+
+    public static void  outOfBoad(){
```
I think this is a typo and you meant outOfBoard? I personally think outOfBounds might be a bit more clear. It sounds like soccer and all the sports ⚽

bushidocodes replied:
Oh and tests are usually a bit more verbose so they say what the expected behavior is in the test. E.g. throwsOutOfBoundExceptionWhenPassedCoordinatesOutOfBounds. Some folks don't like this, but I usually try to be super verbose so my tests and code are easy to understand

bushidocodes commented on src/main/java/edu/gwu/csci6212/MinimumStepsChessKnight.java:
```diff
@@ -0,0 +1,113 @@
+        while(!gameOver){
+            calculateSquares(queue.remove());
+        }
+        System.out.println("The knight can find the goal in " + totalSteps + " steps.");
```
Rather than a comment, I think you want to return the number of steps here so that your test assertions can check this.

bushidocodes commented on src/main/java/edu/gwu/csci6212/MinimumStepsChessKnight.java:
```diff
@@ -0,0 +1,113 @@
+    private int [][] board;
+    private int[] start;
+    private int[] goal;
+    private LinkedList<int[]> queue;
```
A LinkedList. Nice!
![](https://media.giphy.com/media/D7z8JfNANqahW/giphy.gif)

bushidocodes commented on src/main/java/edu/gwu/csci6212/MinimumStepsChessKnight.java:
```diff
@@ -0,0 +1,113 @@
+        this.queue = new LinkedList<>();
+        this.gameOver = false;
+
+        this.knightMovements = new int [][] {
```
Can you explain what the step part of this triple is for? I would have expected a set of ordered pairs. Or perhaps an array of instances of a KnightMove class.

bushidocodes commented on src/main/java/edu/gwu/csci6212/MinimumStepsChessKnight.java:
```diff
@@ -0,0 +1,113 @@
+        this.start[1] = start[1];
+        this.start[2] = 0;
+        this.goal = goal;
+        if (outOfBoard()) throw new Exception("Position out of board");
```
I think "out of bounds" is perhaps more idiomatic English.

bushidocodes commented on src/main/java/edu/gwu/csci6212/MinimumStepsChessKnight.java:
```diff
@@ -0,0 +1,113 @@
+        this.goal = goal;
+        if (outOfBoard()) throw new Exception("Position out of board");
+        board [start[0]][start[1]] = 1;
+        if (!isKnightInGoal()){
```
What happens if the Knight is in the goal. I think we'd want to return 0 since the Knight had to make 0 moves to get to the destination, right?

bushidocodes commented on src/main/java/edu/gwu/csci6212/MinimumStepsChessKnight.java:
```diff
@@ -0,0 +1,113 @@
+
+    private boolean outOfBoard(){
+        if(start[0] > size - 1 || start[0] < 0 || start[1] > size -1 || start[1] < 0) return true;
+        return goal[0] > size - 1 || goal[0] < 0 || goal[1] > size - 1 || goal[1] < 0;
```
It seems like this could use isInsideLimits to DRY up the code a bit.

bushidocodes commented on src/test/java/edu/gwu/csci6212/MinimumStepsChessKnightTest.java:
```diff
@@ -0,0 +1,56 @@
+    public static void executeGame(int size, int[] start, int[] goal){
+        MinimumStepsChessKnight myGame = new MinimumStepsChessKnight();
+        try{
+            myGame.setUpBoard(size, start, goal);
```
I think I'd be surprised that the game would run automatically if I called setUpBoard. Maybe either have setupBoard and play or just rename this function to play.

bushidocodes commented on src/test/java/edu/gwu/csci6212/MinimumStepsChessKnightTest.java:
```diff
@@ -0,0 +1,56 @@
+        try{
+            myGame.setUpBoard(size, start, goal);
+        }catch (Exception e){
+            System.out.println(e);
```
You should probably expose the exception to JUnit tests so that they can test for it. I have an example of that in my code It'd look something like `@Test(expected = IndexOutOfBoundsException.class)`

bushidocodes commented on src/test/java/edu/gwu/csci6212/MinimumStepsChessKnightTest.java:
```diff
@@ -0,0 +1,56 @@
+package edu.gwu.csci6212;
+
+public class MinimumStepsChessKnightTest {
```
As mentioned in WhatsApp, definitely would be nice to see JUnit assertions here.

bushidocodes reviewed (changes requested):
Let me know if you have any questions. I'll need to re-review once these are addressed because I don't yet funny understand the algorithm yet.

albero94 committed:
Create Position instead of using 3 element array.

OutOfBounds function renamed from outOfBoard.

Return -1 value if OutOfBounds start/goal.

Return 0 value if knight is in the goal position.

Return number of steps for the rest of scenarios.

Use isInsideLimits to dry up the code.

Use JUnit assertions for testing Name the test functions in a more verbose way.

Rename setupBoard to startGame

bushidocodes commented on src/main/java/edu/gwu/csci6212/MinimumStepsChessKnight.java:
```diff
@@ -0,0 +1,127 @@
+            this.steps  += p1.steps;
+        }
+        public boolean equals(Position p2){
+            if (this.x == p2.x && this.y == p2.y)   return true;
```
You can dry up lines 56-57 with `return this.x == p2.x && this.y == p2.y`

bushidocodes commented on src/main/java/edu/gwu/csci6212/MinimumStepsChessKnight.java:
```diff
@@ -0,0 +1,127 @@
+    }
+
+    private boolean outOfBounds(){
+        if(!isInsideLimits(start)) return true;
```
Lines 97-99 can be dried up to `return !isInsideLimits(start) || !isInsideLimits(goal)`

albero94 committed:
Dry up return statements.

bushidocodes reviewed (approved):
Looks Good to Me! (LGTM)
